src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -187,26 +198,45 @@ export class WorkspaceSavedObjectsClientWrapper {
     const findWithWorkspacePermissionControl = async <T = unknown>(
       options: SavedObjectsFindOptions
     ) => {
-      if (options.workspaces) {
-        options.workspaces = options.workspaces.filter(
-          async (workspaceId) =>
-            await this.permissionControl.validate(
-              wrapperOptions.request,
-              {
-                type: WORKSPACE_TYPE,
-                id: workspaceId,
-              },
-              [PermissionMode.Read]
-            )
+      const principals = this.permissionControl.getPrincipalsFromRequest(wrapperOptions.request);
+
+      if (this.isRelatedToWorkspace(options.type)) {
+        const queryDSLForQueryingWorkspaces = ACL.genereateGetPermittedSavedObjectsQueryDSL(
+          [PermissionMode.LibraryRead, PermissionMode.LibraryWrite, PermissionMode.Management],
+          principals,
+          WORKSPACE_TYPE
         );
+        options.queryDSL = queryDSLForQueryingWorkspaces;
       } else {
-        options.workspaces = [
-          'public',
-          ...(await this.permissionControl.getPermittedWorkspaceIds(wrapperOptions.request, [
-            PermissionMode.Read,
-          ])),
-        ];
+        const permittedWorkspaceIds = await this.permissionControl.getPermittedWorkspaceIds(
+          wrapperOptions.request,
+          [PermissionMode.LibraryRead, PermissionMode.LibraryWrite, PermissionMode.Management]
+        );
+        if (options.workspaces) {
+          const isEveryWorkspaceIsPermitted = options.workspaces.every((item) =>
+            // TODO modify this line to use permittedWorkspaceIds if public workspace is also a workspace
+            ['public', ...(permittedWorkspaceIds || [])]?.includes(item)
+          );
+          if (!isEveryWorkspaceIsPermitted) {
+            throw generateWorkspacePermissionError();
+          }
```

Comment:
Do we need to add `queryDSL` options here?

Reply:
I do not think we should. Either users have permission on workspace or permission on objects can get access to the objects, and in this if clause, users should access objects under workspace context.